File: androidApp/src/main/java/com/example/kmmweather/android/main/home/HomeViewModel.kt
package com.example.kmmweather.android.main.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.kmmweather.repositories.WeatherRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch

class HomeViewModel(
    private val repository: WeatherRepository
) : ViewModel() {

    var state = MutableStateFlow<HomeViewState>(HomeViewState.NoData)
        private set

    fun requestData(
        latitude: Double,
        longitude: Double,
        forceRemote: Boolean,
    ) {
        viewModelScope.launch {
            val forecast = repository.getForecastForToday(latitude, longitude, forceRemote)
            forecast.collect {
                state.emit(
                    if (it.data == null) HomeViewState.Error(it.error!!.cause)
                    else HomeViewState.Forecast(it.data!!)
                )
            }
        }
    }

    fun requestLocationList() = viewModelScope.launch {
        repository.getForecastList().let {
            state.emit(HomeViewState.ForecastList(it))
        }
    }

}

File: iosApp/iosApp/Views/HomeViewModel.swift
//
//  HomeViewModel.swift
//  iosApp
//
//  Created by Android on 06.04.2023.
//  Copyright © 2023 orgName. All rights reserved.
//
import Foundation
import shared

typealias SharedResult = Result

extension HomeView {
    @MainActor class HomeViewModel: ObservableObject {
        
        let weatherRepository = InjectionHelper().weatherRepository
        
        @Published var currentCityForecast = ForecastBody(
            latitude: 0,
            longitude: 0,
            address: "",
            dateTemperatureRange: "",
            currentHourTemperature: 0,
            weatherDescription: "",
            dayTemperatureString: ""
        )
        @Published var locationList = [Forecast]()
        @Published var errorMessage: String? = nil
        
        lazy var vmCollector: Observer = {
            let collector = Observer { value in
                if let value = value as? SharedResult<ForecastBody> {
                    let data = value
                    self.currentCityForecast = data.data ?? self.currentCityForecast
                    self.errorMessage = data.error?.cause
                }
            }
            return collector
        }()
        
        func requestLocation(lat: Double, lon: Double, forceRemote: Bool) {
            DispatchQueue.main.async {
                self.weatherRepository.getForecastForToday(latitude: lat, longitude: lon, forceRemote: forceRemote, completionHandler: { data, error in
                    data?.collect(collector: self.vmCollector, completionHandler: { err in
                        if (err != nil) {
                            self.errorMessage = err?.localizedDescription
                        }
                    })
                })
            }
        }
        
        func requestLocationList() {
            DispatchQueue.main.async {
                self.weatherRepository.getForecastList(completionHandler: { data, error in
                    if (data != nil) {
                        self.locationList = data!
                    }
                })
            }
        }
    }
}
